### ISSUE TITLE
RealmList.contains() performance

### ISSUE THREAD

**Issue by decnorton** [realm/realm-java#897]
Hi guys,

To get around the fact that relationships allow the same object twice I have to check to see if the object I want to add is already in the RealmList. Currently `contains()` falls back to the `AbstractCollection` implementation that iterates through each object and calls `.equals()`, which is really slow. I changed this to run a query instead and got much better results:

```
No check:
V/RealmRepository﹕ ⇢ ensureRelations()
V/RealmRepository﹕ ⇠ ensureRelations [1295ms]

With AbstractCollection.contains():
V/RealmRepository﹕ ⇢ ensureRelations()
V/RealmRepository﹕ ⇠ ensureRelations [45773ms]

With query: RealmList.where().equalTo("id", id).count() == 0
V/RealmRepository﹕ ⇢ ensureRelations()
V/RealmRepository﹕ ⇠ ensureRelations [3196ms]
```

Of course the query only works if you know the colum name and value, but perhaps primary keys could be used for this? I'm not very familiar with the Realm internals so I don't know if there's a better way of checking if an object exists in a view!

**Comment by emanuelez** [comment]
You are right. We have to override RealmList.contains(). Stay tuned! :)

**Comment by cmelchior** [comment]
The LinkView class in C++ has a `find` method:

```
    /// Search this list for a link to the specified target table row (specified
    /// by its index in the target table). If found, the index of the link to
    /// that row within this list is returned, otherwise `realm::not_found` is
    /// returned.
    size_t find(size_t target_row_ndx, size_t start=0) const noexcept;
```

This could be implemented like this:

1) Add `nativeFind(long linkViewPtr, long tableRowIndex)` to `io.realm.internal.LinkView` and the JNI layer in `io_realm_internal_LinkView.cpp`.

2) add a `contains(long tableRowIndex)` method to `io.realm.internal.LinkView`

3) RealmList.contains()
```
if (!sameRealm) {
  return false
} else {
  return linkView.contains(tableRowIndex)
}
```

**Comment by donnfelker** [comment]
I implemented this, but when using a index, if we have two objects from two separate Realm's that could introduce falst positives. Thats what my tests are showing here: https://github.com/donnfelker/realm-java/commit/79b644c439c5f2bf710f7acf10420ec2b34eb497#diff-366f9ca7ccea0ef55d654a1132aeb49bR729  Using a row index might be a bad decision here. Are there any other options in core?

**Comment by cmelchior** [comment]
We should just check that the Realm is the proper type before calling into LinkView, that is what e.g `equals` is doing:

```
        String path = realm.getPath();
        String otherPath = aAllTypes.realm.getPath();
        if (path != null ? !path.equals(otherPath) : otherPath != null) return false;;
```

**Comment by cmelchior** [comment]
This has been fixed by #1901 and has been released as part of 0.87.2
